feat: add user registration troubleshooting how-to guide (#100)
canonical/charmed-aether-sd-core: docs/how-to/troubleshoot_user_registration_issues.md

# Troubleshoot subscriber registration issues
This guide provides step-by-step troubleshooting actions to remediate subscriber registration issues. We hope you don't need this guide. If you encounter an issue and aren't able to address it via this guide, please raise an issue [here][Bug Report].

The examples use GNBSIM as the 5G gNodeB and UE simulator.

## 1. Subscriber registration failure

### Symptoms
The `juju debug-log sdcore-gnbsim-k8s/leader` command reports an error in the UE registration:
```console
handling event: REGISTRATION-REJECT-EVENT	{"component": "GNBSIM", "category": "SimUe", "supi": "imsi-001010100007487"}
```
#### Possible reasons
This error is typically raised when the UE is not provisioned in SD-Core or the link between AMF and AUSF or between AUSF and UDM/UDR is not established.

### Recommended Actions
#### Provisioning

> Validate that the UE is provisioned in SD-Core, either via the UI or the API. Validate that the UE trying to connect to the core has the same IMSI.

Reviewer: what does it mean to be "provisioned in the SD-Core"? I think we should point what to look at in the NMS.

The command should report `Running` for all the four pods.
```console
NAME                            READY   STATUS    RESTARTS   AGE
nrf-0                           2/2     Running   0          15m
ausf-0                          2/2     Running   0          15m
udm-0                           2/2     Running   0          13m
udr-0                           2/2     Running   0          13m
```
If any of the pods is not in `Running` state, you can try to restart it using `kubectl kill <pod-name> -n sdcore`.

## 2. Subscriber authentication failure

### Symptoms
The `juju debug-log sdcore-gnbsim-k8s/leader` command reports an error in the UE authentication:
```console
handling event: AUTHENTICATION-REJECT-EVENT	{"component": "GNBSIM", "category": "SimUe", "supi": "imsi-001010100007487"}
```
#### Possible reasons
This error is typically raised when the UE is provisioned in SD-Core but the authentication (OPc and/or key) data in SD-Core does not match the actual UE data or the link between AMF and AUSF or between AUSF and UDM/UDR is not established.

### Recommended Actions
#### Provisioning

> Validate whether the UE authentication in SD-Core matches the actual UE authentication data, either via the UI or the API.

Reviewer: "UE authentication in SD-Core" -> where do I see those values?
"Actual UE Auth data" -> after reading the  possible reasons part, I suppose that's the OPC/Key. So I would point to where exactly see those fields in the NMS

#### Configuration
If the logs do not match, ensure that NRF, AUSF, UDM, UDR pods are up and running using `kubectl get pods -n sdcore`.

> The command should report `Running` for all the three pods.

Reviewer 2: ```suggestion
The command should report `Running` for all the four pods.
```
We listed four pods.

```console
NAME                            READY   STATUS    RESTARTS   AGE
nrf-0                           2/2     Running   0          15m
ausf-0                          2/2     Running   0          15m
udm-0                           2/2     Running   0          13m
udr-0                           2/2     Running   0          13m
```
If any of the pods is not in `Running` state, you can try to restart it using `kubectl kill <pod-name> -n sdcore`.

## 3. Session establishment failure

### Symptoms

The `juju debug-log sdcore-gnbsim-k8s/leader` command reports an error in the UE authentication:
```console
handling event: PDU-SESSION-ESTABLISHMENT-REJECT-EVENT	{"component": "GNBSIM", "category": "SimUe", "supi": "imsi-001010100007487"}
```
#### Possible reasons
This error is typically raised when the UE is provisioned in SD-Core but the DNN (Data Network Name) and or the S-NSSAI requested by the UE are not available in SD-Core. It is also possible that the AMF is not able to select an SMF, or the SMF is not able to connect to UPF.

### Recommended Actions
#### Provisioning

> Validate whether the DNN/S-NSSAI requested by the UE are available in the configured Network Slice and Device Group for the UE, either via the UI or the API.

Reviewer: where do I get the DNN/S-NSSAI requested by the UE and where do I see the values used by the NS/DG (what fields do I need to look at) ? 
We cannot see the DNN or S-NSSAI using the NMS.

INFO	handler/handler.go:530	handle PFCP Session Modification Response	{"component": "SMF", "category": "PFCP"}
INFO	handler/handler.go:567	PFCP Modification Response Accept	{"component": "SMF", "category": "PduSess", "uuid": "urn:uuid:554a3fbb-0d23-451b-b3e9-c6fad3e96c8c", "id": "imsi-001010100007487", "pduid": 10}
INFO	handler/handler.go:588	PFCP Session Modification Success[4]	{"component": "SMF", "category": "PFCP", "uuid": "urn:uuid:554a3fbb-0d23-451b-b3e9-c6fad3e96c8c", "id": "imsi-001010100007487", "pduid": 10}
INFO	context/sm_context.go:276	context state change, current state[SmStatePfcpModify] next state[SmStateActive]	{"component": "SMF", "category": "CTX", "uuid": "urn:uuid:554a3fbb-0d23-451b-b3e9-c6fad3e96c8c", "id": "imsi-001010100007487", "pduid": 10}
INFO	context/sm_context.go:276	context state change, current state[SmStateActive] next state[SmStateActive]	{"component": "SMF", "category": "CTX", "uuid": "urn:uuid:554a3fbb-0d23-451b-b3e9-c6fad3e96c8c", "id": "imsi-001010100007487", "pduid": 10}
```
If the logs do not match, ensure that NRF, AUSF, SMF, UPF pods are up and running using `kubectl get pods -n sdcore`.

> The command should report `Running` for all the three pods.

Reviewer 2: We have more than 3 pods here.

# Troubleshoot subscriber registration issues

> This guide provides step-by-step troubleshooting actions to remediate subscriber registration issues issues. We hope you don't need this guide. If you encounter an issue and aren't able to address it via this guide, please raise an issue [here][Bug Report].

Reviewer 3: ```suggestion
This guide provides step-by-step troubleshooting actions to remediate subscriber registration issues issues. We hope you don't need this guide. If you encounter an issue and aren't able to address it via this guide, please raise an issue [here][Bug Report]. The examples here use GNBSIM as the 5G gNodeB and UE simulator. 
```

> ## 1. Subscriber registration failed with `REGISTRATION-REJECT-EVENT` error

Reviewer 3: ```suggestion
## 1. Subscriber registration failed
```

This log comes from gnbsim. Most users connecting to the core won't do it with gnbsim (but with an actual radio and UE)

### Symptoms
The `juju debug-log sdcore-gnbsim-k8s/leader` command reports an error in the UE registration:
```console
handling event: REGISTRATION-REJECT-EVENT	{"component": "GNBSIM", "category": "SimUe", "supi": "imsi-001010100007487"}
```

### Recommended Actions

> This error is typically raised when the UE is not provisioned in SD-Core or the link between AMF and AUSF or between AUSF and UDM/UDR is not established.

Reviewer 3: This sentence is not a "Recommended Action" but an explanation, if those are needed, please add a section for it.

#### Provisioning

> Validate whether the UE is provisioned in SD-Core.

Reviewer 3: ```suggestion
Validate that the UE is provisioned in SD-Core, either via the UI or the API. Validate that the UE trying to connect to the core has the same IMSI.
```

INFO	gmm/handler.go:2254	Handle Registration Complete	{"component": "AMF", "category": "GMM", "amf_ue_ngap_id": "AMF_UE_NGAP_ID:2", "suci": "suci-0-001-01-0-0-0-0100007487", "supi": "SUPI:imsi-001010100007487"}
INFO	fsm/fsm.go:99	handle event[ContextSetup Success], transition from [ContextSetup] to [Registered]	{"component": "LIB", "category": "FSM"}
```

Validate the AUSF is able to select an UDM by inspecting logs using `microk8s.kubectl logs -n sdcore ausf-0 -c ausf -f`. The logs should report:
```console
INFO	producer/ue_authentication.go:84	HandleUeAuthPostRequest	{"component": "AUSF", "category": "UeAuthPost"}
INFO	producer/ue_authentication.go:66	Auth5gAkaComfirmRequest	{"component": "AUSF", "category": "5gAkaAuth"}
```

Validate the UDM is able to select an UDR by inspecting logs using `microk8s.kubectl logs -n sdcore udm-0 -c udm -f`. The logs should report no errors after this line:
```console
INFO	producer/generate_auth_data.go:84	handle GenerateAuthDataRequest	{"component": "UDM", "category": "UEAU"}
```

> ## 2. Subscriber authentication failed with `AUTHENTICATION-REJECT-EVENT` error

Reviewer 3: ```suggestion
## 2. Subscriber authentication failed
```

### Symptoms
The `juju debug-log sdcore-gnbsim-k8s/leader` command reports an error in the UE authentication:
```console
handling event: AUTHENTICATION-REJECT-EVENT	{"component": "GNBSIM", "category": "SimUe", "supi": "imsi-001010100007487"}
```

### Recommended Actions

> This error is typically raised when the UE is provisioned in SD-Core but the authentication (OPc and/or key) data in SD-Core does not match the actual UE data or the link between AMF and AUSF or between AUSF and UDM/UDR is not established.

Reviewer 3: This sentence is not a "Recommended Action" but an explanation, if those are needed, please add a section for it.

#### Provisioning

> Validate whether the UE authentication in SD-Core matches the actual UE authentication data.

Reviewer 3: ```suggestion
Validate whether the UE authentication in SD-Core matches the actual UE authentication data, either via the UI or the API.
```

#### Configuration
See "Configuration" in previous section.

> ## 3. Session establishment failed with `PDU-SESSION-ESTABLISHMENT-REJECT-EVENT` error

Reviewer 3: ```suggestion
## 3. Session establishment failed
```

### Symptoms

The `juju debug-log sdcore-gnbsim-k8s/leader` command reports an error in the UE authentication:
```console
handling event: PDU-SESSION-ESTABLISHMENT-REJECT-EVENT	{"component": "GNBSIM", "category": "SimUe", "supi": "imsi-001010100007487"}
```

### Recommended Actions

> This error is typically raised when the UE is provisioned in SD-Core but the DNN (Data Network Name) and or the S-NSSAI requested by the UE are not available in SD-Core. It is also possible that the AMF is not able to select an SMF, or the SMF is not able to connect to UPF.

Reviewer 3: Same comment around the fact that this block does not fit well in this section.

#### Provisioning

> Validate whether the DNN/S-NSSAI requested by the UE are available in the configured Network Slice and Device Group for the UE.

Reviewer 3: ```suggestion
Validate whether the DNN/S-NSSAI requested by the UE are available in the configured Network Slice and Device Group for the UE, either via the UI or the API.
```
